packages/kas/src/parser.ts

```diff
@@ -0,0 +1,387 @@
+
+    subscriptable() {
+        const index = this.index; // save state
+        try {
+            const left = this.variable();
+            if (this.peek().kind === "_") {
+                this.consume();
+                const right = this.subscriptable();
+                return new Var(left, right);
+            } else {
+                return new Var(left);
+            }
+        } catch (e) {
+            this.index = index; // restore state
+            const next = this.peek();
+            this.consume();
+            switch (next.kind) {
+                case "CONST":
+                    return new Const(next.value.toLocaleLowerCase());
+                case "INT":
+                    // Why are we using the `Number` constructor instead of `parseInt`?
+                    return Int.create(Number(next.value));
+                case "FLOAT":
+                    // Why are we using the `Number` constructor instead of `parseFlat`?
+                    return Float.create(Number(next.value));
+                case "{": {
+                    const node = this.additive().completeParse(); // post-process Trig ndoes with exponents
+                    this.expect("}");
+                    return node;
+                }
+                case "(": {
+                    const node = this.additive()
+                        .completeParse() // post-process Trig ndoes with exponents
+                        .addHint("parens"); // this probably shouldn't be a hint...
+                    this.expect(")");
+                    return node;
+                }
+                default:
+                    throw new Error(
+                        `Expected either CONST, INT, FLOAT, {, or ( but got ${next.kind}`,
+                    );
+            }
+        }
+    }
+
+    function() {
+        const next = this.peek();
+        if (next.kind === "FUNC") {
+            this.consume();
+            return next.value;
+        } else {
+            throw new Error(`Expected FUNC token but got ${next.kind}`);
+        }
+    }
+
+    invocation() {
+        const next = this.peek();
+        this.consume();
+        switch (next.kind) {
+            case "sqrt": {
+                const next = this.peek();
+                this.consume();
+                switch (next.kind) {
+                    case "(": {
+                        const arg = this.additive();
+                        this.expect(")");
+                        return Pow.sqrt(arg);
+                    }
+                    case "{": {
+                        const arg = this.additive();
+                        this.expect("}");
+                        return Pow.sqrt(arg);
+                    }
+                    case "[": {
+                        const index = this.additive();
+                        this.expect("]");
+                        this.expect("{");
+                        const arg = this.additive();
+                        this.expect("}");
```

Comment:
TODO: encapsulate this pattern in a helper function called `between()`.  Calling it will look something like this:
```ts
this.between("{", "additive", "}");
```